src/azul/drs.py

```diff
@@ -74,10 +87,139 @@ class Access:
     headers: Optional[Mapping[str, str]] = None
 
 
+class DRSURI(metaclass=ABCMeta):
+
+    @classmethod
+    def parse(cls, drs_uri: str) -> 'DRSURI':
+        scheme = 'drs://'
+        require(drs_uri.startswith(scheme))
+        # "The colon character is not allowed in a hostname-based DRS URI".
+        # https://ga4gh.github.io/data-repository-service-schemas/preview/develop/docs/#_drs_uris
+        subcls = CompactDRSURI if drs_uri.find(':', len(scheme)) >= 0 else RegularDRSURI
+        return subcls.parse(drs_uri)
+
+    @abstractmethod
+    def to_url(self, client: 'DRSClient', access_id: Optional[str] = None) -> str:
+        """
+        Translate the DRS URI into a DRS URL. All query params included in the DRS
+        URI (eg '{drs_uri}?version=123') will be carried over to the DRS URL.
+        """
+        raise NotImplementedError
+
+
+@attr.s(auto_attribs=True, kw_only=True, frozen=True, slots=True)
+class RegularDRSURI(DRSURI):
+    uri: furl
+
+    def __attrs_post_init__(self):
+        assert self.uri.scheme == 'drs', self.uri
+
+    @property
+    def netloc(self) -> str:
+        return self.uri.netloc
+
+    @classmethod
+    def parse(cls, drs_uri: str) -> 'RegularDRSURI':
+        return cls(uri=furl(drs_uri))
+
+    def to_url(self, client: 'DRSClient', access_id: Optional[str] = None) -> str:
+        url = self.uri.copy().set(scheme='https')
+        url.set(path=drs_object_url_path(object_id=one(self.uri.path.segments),
+                                         access_id=access_id))
+        return str(url)
+
+
+@attr.s(auto_attribs=True, kw_only=True, frozen=True, slots=True)
+class CompactDRSURI(DRSURI):
+    """
+    So-called DRS "URIs" [1] for Compact Identifiers [2] are NOT URIs according
+    to RFC 3986 [3] so we can't use off-the-shelf URI parsers.
+
+    [1] https://ga4gh.github.io/data-repository-service-schemas/preview/release/drs-1.3.0/docs/
+
+    [2] https://www.nature.com/articles/sdata201829
+
+    [3] https://datatracker.ietf.org/doc/html/rfc3986
+    """
+    _namespace: str
+    _accession: str
```

Comment:
```suggestion
    namespace: str
    accession: str
```

This was in my patch but I changed my mind. I now see no need to protect these because they are well-documented parts of the spec.